Implement Form Validation

Sunny-xyz opened the pull request:
This PR introduces form validations across the Quiz app. In the models, validations have been added to both the Quiz and Question classes so that essential fields (such as title, description, question content, answer options, and the correct option) must be present before records are saved. This was done to improve data integrity by preventing invalid data from being stored in the database.

On the view side, forms have been updated to incorporate GOV.UK error components, providing the user with immediate feedback when validation fails. Error summary alerts are displayed at the top of the forms to notify the user of issues, while additional error messages appear next to each field with errors. Moreover, user input is retained on form re-rendering, so that users don’t lose the data they have already entered.

Sunny-xyz committed:
Implemented Form Validation and Error Handling

This commit introduces form validations across the Quiz app. In the models, validations have been added to both the Quiz and Question classes so that essential fields (such as title, description, question content, answer options, and the correct option) must be present before records are saved. This was done to improve data integrity by preventing invalid data from being stored in the database.

On the view side, forms have been updated to incorporate GOV.UK error components, providing the user with immediate feedback when validation fails. Error summary alerts are displayed at the top of the forms to notify the user of issues, while additional error messages appear next to each field with errors. Moreover, user input is retained on form re-rendering, so that users don’t lose the data they have already entered.

After receiving some feedback from Kev, I've made some changes to this commit in a rebase; I've further refined the error rendering by using the input component's error_items attribute. Instead of manually rendering error messages, model errors are now passed directly into the components, so that the validation messages defined in the model are automatically displayed.

Sunny-xyz committed:
Ensured update action re-renders edit view on validation failures

Previously, when a user submitted invalid fields during question editing, the update action would simply do nothing—leaving the user stuck on a blank page without any error feedback. To address this, I've updated the update action in the QuestionsController so that it now re-renders the edit view with an Unprocessable Entity status when validations fail. For consistency, I’ve also updated the Create and Update actions in QuizzesController to re-render the view with the same status upon failure.

Sunny-xyz committed:
Added rspec tests for quizzes and questions validations

This commit introduces additional request specs for both quizzes and questions to verify that invalid form submissions are handled correctly.

For quizzes, the new tests check that attempts to create or update a quiz with missing required fields return an Unprocessable Entity status, and that the re-rendered form displays an error alert along with error messages (e.g. "Please enter a title for your quiz" and "Please provide a description for your quiz").

Similarly, for questions, new tests have been added to ensure that invalid parameters during both the creation and update processes result in error messages being displayed and that no invalid records are created.

kevindew commented on app/controllers/questions_controller.rb:
```diff
@@ -11,7 +11,7 @@ def create
       redirect_to quiz_path(@quiz)
     else
-      render :new
+      render :new, status: :unprocessable_entity
```
Nice one adding this status code 👍

Sunny-xyz replied:
Thanks!

kevindew commented on app/views/questions/_form.html.erb:
```diff
@@ -1,38 +1,74 @@
     value: @question.content,
     id: "question_content"
   } %>
+  <% if @question.errors[:content].any? %>
+    <%= render "govuk_publishing_components/components/error_message", {
+      text: "Question can't be blank",
+      id: "question_content_error"
+    } %>
+  <% end %>
```
So there's a couple of things that can be improved with this.

We expect to use the error message from the validator itself and not provide it here (if we want to customise it we do it at the validator: https://guides.rubyonrails.org/active_record_validations.html#examples)

We also can add errors directly into the input component: https://components.publishing.service.gov.uk/component-guide/input/with_error_items

So what I'd suggest doing is passing through the errors from the model:

```
error_items: question.errors[:content].map { |message| { text: message } }
```

(In GOV.UK Chat we've got a helper to do this for us: https://github.com/alphagov/govuk-chat/blob/16d27a1d821243d768e266b7c7235925808c23a4/app/helpers/errors_helper.rb#L9-L11 used like so: https://github.com/alphagov/govuk-chat/blob/16d27a1d821243d768e266b7c7235925808c23a4/app/views/sign_up/found_chat.html.erb#L20 which you may want to do to avoid frequent code repetition)

Sunny-xyz replied:
Thanks for the feedback on this along with the suggestions - I'll look into adding a helper function for the errors and see what I can do!

kevindew commented on spec/requests/questions_spec.rb:
```diff
@@ -81,6 +101,34 @@
+      patch quiz_question_path(quiz, question), params: { question: {
+        content: "",
+        option_a: "",
+        option_b: "Answer B",
+        option_c: "Answer C",
+        option_d: "Answer D",
+        correct_option: "option_a",
```
```suggestion
```

Since this is a patch request you shouldn't need to pass in all the attributes (Patch requests are able to update just a subset of attributes)

Sunny-xyz replied:
Ahh, gotcha, I'll get this changed for the next commit

kevindew commented on app/views/questions/_form.html.erb:
```diff
@@ -1,38 +1,74 @@
+  <%= render "govuk_publishing_components/components/error_alert", {
+    message: "There were errors with your submission. Please fix them below."
+  } %>
+<% end %>
```
Not something you need to do on this PR but something I thought I should flag.

On GOV.UK services we have a pattern of listing validation errors at the top of a page as well as on the individual field. It's explained somewhat in: https://design-system.service.gov.uk/components/error-summary/.

You can see how in GOV.UK Chat we've implemented this in: https://github.com/alphagov/govuk-chat/blob/16d27a1d821243d768e266b7c7235925808c23a4/app/views/sign_up/found_chat.html.erb#L7-L9 with the helper: https://github.com/alphagov/govuk-chat/blob/16d27a1d821243d768e266b7c7235925808c23a4/app/helpers/errors_helper.rb#L2-L7

Sunny-xyz replied:
I see - thanks for this Kev, I'll look into applying this to the app ASAP

Sunny-xyz replied:
I think I'll create a separate PR for this, but it i'll definitely be good to do next

kevindew reviewed (commented):
Great job on the commit messages here Sunny ⭐ 

So I think the main thing to look at here is aspects of the first commit where we'll want to use validation error messages from the model itself and then pushing the errors into the input fields.

Only other thing I'd feedback is that it'd be useful for a PR like this to have screenshots of the validation occurring so I can see how it looks visually without running the code.

Sunny-xyz commented:
> Great job on the commit messages here Sunny ⭐
> 
> So I think the main thing to look at here is aspects of the first commit where we'll want to use validation error messages from the model itself and then pushing the errors into the input fields.
> 
> Only other thing I'd feedback is that it'd be useful for a PR like this to have screenshots of the validation occurring so I can see how it looks visually without running the code.

Thanks a ton for all the feedback Kev! I'll have time later today to work on this PR some more (currently tied up in the jailbreak guardrails) so I'll push the appropriate changes then

kevindew commented:
Thanks @Sunny-xyz no rush - feel free to prioritise the guardrails where there are clear tasks to do there.

kevindew commented on app/views/quizzes/edit.html.erb:
```diff
@@ -1,20 +1,38 @@
+      text: "Title can't be blank",
+      id: "quiz_title_error"
+    } %>
+  <% end %>
```
This still has the hardcoded error approach.

I imagine you're at a point where it'd be better to combine edit.html.erb and new.html.erb into a single _form.html.erb form partial so you don't have to repeat the same code twice.

Sunny-xyz replied:
Cool, I'll get this done ASAP. Would you recommend doing it in this PR or would that start to fall outside of the scope of this one?

kevindew commented on spec/requests/questions_spec.rb:
```diff
@@ -81,6 +101,29 @@
+      } }
+      expect(response).to have_http_status(:unprocessable_entity)
+      expect(response.body).to include("There were errors with your submission. Please fix them below.")
+      expect(response.body.scan("Field can&#39;t be blank").size).to be >= 2
```
This seems a bit cryptic - I'm assuming you're wanting to check that this appears twice so I'm not sure why we're asserting that is potentially greater than 2 times?

Sunny-xyz replied:
This was done just in case we wanted to modify the properties of the underlying quiz to have more failures, but I think for now it would be better to just keep it at ' = 2'; I'll change this now

kevindew commented on spec/requests/quizzes_spec.rb:
```diff
@@ -48,6 +48,19 @@
+        post quizzes_path, params: { quiz: { title: "", description: "" } }
+      }.not_to change(Quiz, :count)
+
+      expect(response).to have_http_status(:success)
```
We'd want this to be unprocessable_entity as per the question ones.

Sunny-xyz replied:
Got it, thanks!

kevindew commented on spec/requests/quizzes_spec.rb:
```diff
@@ -63,6 +76,18 @@
+
+    it "does not update the quiz and displays error messages" do
+      patch quiz_path(quiz.id), params: { quiz: { title: "", description: "" } }
+      expect(response).to have_http_status(:success)
```
Similar to above about this being unprocessable_entity

Sunny-xyz replied:
Changing now

kevindew commented on spec/requests/quizzes_spec.rb:
```diff
@@ -63,6 +76,18 @@
+      patch quiz_path(quiz.id), params: { quiz: { title: "", description: "" } }
+      expect(response).to have_http_status(:success)
+      expect(response.body).to include("There were errors with your submission. Please fix them below.")
+      expect(response.body).to include("Title can&#39;t be blank")
+      expect(response.body).to include("Description can&#39;t be blank")
```
These are peculiar - they don't match the validation messages you've configured in your model - are we sure this is being run?

Sunny-xyz replied:
That's really odd. I'll look into this now. I didn't catch it due the test itself being successful, but it shouldn't have been.

Sunny-xyz replied:
Ah, right, this makes sense now actually - as per the suggestion you left above, the edit view still has the hard-coded error messages "Title can't be blank" and "Description can't be blank". Changing those makes this test fail, as I've confirmed now

kevindew replied:
Oh yes I'd forgot about that - nice one

kevindew reviewed (commented):
Thanks, added a few more comments of things to look at a bit more.

Sunny-xyz committed:
Combined Quiz Forms into Shared Partial

This commit extracts the common quiz form code from the new and edit views into a single shared partial (_form.html.erb). The partial accepts a local variable for the button text, ensuring that "Create Quiz" and "Update Quiz" can be set appropriately in new and edit contexts respectively.

This refactoring reduces code duplication, ensures consistent error handling and styling across quiz forms, and simplifies future changes.

kevindew reviewed (approved):
Nice one, looks good to me.
